src/basilisp/core.lpy

```diff
@@ -3258,6 +3258,21 @@
              (transient {})
              maps))))
 
+(defn merge-with
+  "Merge ``maps`` together from left to right, while reducing values
+  under the same key by successively applying function ``f`` on the
+  preceding and subsequent key values, i.e.
+
+  (``f`` (``f`` (``f`` first-value second-value) third-value) ...)."
+  [f & maps]
+  (reduce (fn [acc kv]
+            (let [k (first  kv)
+                  v (second kv)]
+              (if (contains? acc k)
+                (update acc k f v)
+                (assoc acc k v))))
+          (first maps) (apply concat (rest maps))))
```

Comment:
```suggestion
  (when (some identity maps)
    (persistent!
      (reduce (fn [acc kv]
                (let [k (first  kv)
                      v (second kv)]
                  (if (contains? acc k)
                    (assoc! acc k (f (get acc v) v))
                    (assoc! acc k v))))
              (transient (first maps))
              (apply concat (rest maps))))
```

The main merge uses transients for performance optimization. I've proposed the same here. If it doesn't work, we can return to the persistent-only implementation.

I think we should leave the `when` check in both cases.

Reply:
sorry this didn't work

```
FAIL in (merge-with-test) (test_core_fns.lpy:1102)
    Test failure: (= {:d 7 :b "235" :z 8 :c "46" :a 0} (merge-with str {:b 2 :a 0} {:b 3 :z 8 :c 4} {:d 7 :b 5 :c 6}))

    expected: {:d 7 :b "235" :z 8 :c "46" :a 0}
      actual: {:d 7 :b "5" :z 8 :c "6" :a 0}

FAIL in (merge-with-test) (test_core_fns.lpy:1103)
    Test failure: (= {:b [2 5] :c 7 :a 0} (merge-with conj {:b [2] :a 0} nil {:b 5 :c 7}))

    expected: {:b [2 5] :c 7 :a 0}
      actual: {:b (5) :c 7 :a 0}
```

Reply:
Clojure's version doesn't use transients either, so maybe there's a reason. Thanks for adding the `when` check at least.